Add flux and numerical flux computation for ScalarWave

nilsdeppe opened the pull request:
## Proposed changes

- Add flux computation for the scalar wave system
- Add numerical flux computation for scalar wave system. The numerical flux is computed in two steps. First the `package_data` function is called to compute the contribution from one side of the interface. The call operator of the class is passed the data from each side of the interface and then computes the numerical fluxes.
- Add compute largest characteristic
- Add `doc-check` target to check warnings in documentation. Fix #255 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Add ScalarWave::ComputeNormalDotFluxes

nilsdeppe committed:
Add ScalarWave::UpwindFlux<Dim> numerical flux

osheamonn commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +27,87 @@ void ComputeDuDt<Dim>::apply(
+              psi_normal_dot_flux->get().end(), 0.0);
+  }
+
+  get(*pi_normal_dot_flux) = interface_unit_normal.get(0) * phi.get(0);
```
I think we said we'd prefer get<0>(interface_unit_normal)?

osheamonn commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +27,87 @@ void ComputeDuDt<Dim>::apply(
+
+template <size_t Dim>
+void ComputeNormalDotFluxes<Dim>::apply(
+    gsl::not_null<Scalar<DataVector>*> pi_normal_dot_flux,
```
const for the not_null's

osheamonn commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -68,4 +69,97 @@ struct ComputeDuDt {
+  static constexpr OptionString help = {"Class help text"};
+
+  // clang-tidy: non-const reference
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
```
Why does this need to be pupable?

nilsdeppe replied:
Anything created from options needs to be pupable

wthrowe replied:
DomainCreators aren't pupable.  No idea if this class has to be, but creatability doesn't seem to imply it.

nilsdeppe replied:
True, this is stored in the global cache, though and aren't those created on Main then sent? (looking now)

nilsdeppe replied:
Yea, it looks like everything in the ConstGlobalCache is serializable: https://github.com/sxs-collaboration/spectre/blob/cc563ba35b8b1f55ae8bce5470e551ed525655f4/src/Parallel/Main.hpp#L197 or have I misunderstood that?

wthrowe replied:
Ah, yeah, global cache requiring serializability sounds right.

nilsdeppe replied:
Okay, cool! Since Will and I both weren't sure and are the people most familiar with that code, I'll add a comment explaining why the pup function is there.

osheamonn commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -68,4 +69,97 @@ struct ComputeDuDt {
+
+  using slice_tags = tmpl::list<Pi>;
+
+  void package_data(
```
Could or should this be a static function? I also think it should be documented.

nilsdeppe replied:
In this particular case, yes, in general, no.

Edit:

The reason is that some fluxes have parameters, and as we move to increasingly generic fluxes I wouldn't be surprised if those parameters were needed inside the `package_data` function not just in the call operator.

osheamonn commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -83,3 +83,149 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.DuDt",
+
+template <size_t Dim>
+void check_upwind_flux(const size_t npts, const double t) {
+  ScalarWave::Solutions::PlaneWave<Dim> solution(
```
const

osheamonn commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -83,3 +83,149 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.DuDt",
+      make_array<Dim>(0.1), make_array<Dim>(0.0),
+      std::make_unique<MathFunctions::Gaussian>(1.0, 1.0, 0.0));
+
+  tnsr::I<DataVector, Dim> x = [npts]() {
```
const?

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -68,4 +68,29 @@ struct ComputeDuDt {
+  using return_tags =
+      typelist<Tags::NormalDotFlux<Pi>, Tags::NormalDotFlux<Phi<Dim>>,
+               Tags::NormalDotFlux<Psi>>;
+  using argument_tags = tmpl::list<Pi, Phi<Dim>>;
```
make typelist and tmpl::list consistent

nilsdeppe commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -83,3 +83,64 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.DuDt",
   check_du_dt<3>(3, time);
 }
+
+namespace {
```
const more things

nilsdeppe commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -83,3 +83,64 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.DuDt",
+  const auto pi = Scalar<DataVector>{-get(solution.dpsi_dt(x, t))};
+  const auto phi = solution.dpsi_dx(x, t);
+
+  // Any normal is fine, doesn't even need to be a unit
```
"Any numbers are fine, doesn't have anything to do with unit normal"

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
+ *                   + (n_i)_{\mathrm{int}} F(\Pi)_{\mathrm{int}}
+ *                   - (n_i)_{\mathrm{ext}} F(\Pi)_{\mathrm{ext}}\right)
+ * \f}
+ * where $F$ is the flux computed in ScalarWave::ComputeNormalDotFluxes
```
\f$

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
+ private:
+  struct NormalTimesFluxPi {
+    using type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+    static constexpr const char* const label = "UpwindFlux";
```
db::DataBoxString

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
+
+ public:
+  using options = tmpl::list<>;
+  static constexpr OptionString help = {"Class help text"};
```
Write the help

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
+
+  using slice_tags = tmpl::list<Pi>;
+
+  void package_data(
```
`// pseudo-interface: used internally by Algorithm infrastructure, not user-level code`

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -93,4 +94,72 @@ struct ComputeNormalDotFluxes {
+      const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+      const noexcept;
+
+  void operator()(
```
`// pseudo-interface: used internally by Algorithm infrastructure, not user-level code`

nilsdeppe commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -54,6 +54,60 @@ void ComputeNormalDotFluxes<Dim>::apply(
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+    const noexcept {
+  get(get<Pi>(*packaged_data)) = 0.5 * get(pi);
```
add comment describing math done on one side, as well as note the factor of 1/2

nilsdeppe commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -144,3 +144,88 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.NormalDotFluxes",
+void apply_numerical_flux(
+    const FluxType& flux,
+    const Variables<tmpl::list<Tags...>>& packaged_data_int,
+    const Variables<tmpl::list<Tags...>>& packaged_data_ext, Args&&... args) {
```
remove args

nilsdeppe commented on tests/Unit/Evolution/Systems/ScalarWave/Test_Equations.cpp:
```diff
@@ -144,3 +144,88 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.NormalDotFluxes",
+    return coords;
+  }();
+
+  // Any normal is fine, doesn't even need to be a unit
```
update comment here just as in NormalDotFluxes

nilsdeppe committed:
Fix const-correctness in ScalarWave::ComuteDuDt

nilsdeppe committed:
Add ScalarWave::ComputeLargestCharacteristicSpeed

nilsdeppe commented:
I've rebased this on develop and added 1 new commit, computing the largest characteristic. This commit won't really fit in any of my other PRs so I figured I'd add it here.

@wthrowe @osheamonn @kidder I think this is ready (will need a squash rebase)

[Documentation](http://nilsdeppe.com/spectre/pr_476/)

wthrowe commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -171,4 +171,11 @@ struct UpwindFlux {
+
+/// Compute the maximum magniture of the characteristic speeds.
+struct ComputeLargestCharacteristicSpeed {
+  using return_tags = tmpl::list<>;
```
Why can this return things?

kidder commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -171,4 +171,11 @@ struct UpwindFlux {
           normal_times_flux_pi_exterior) const noexcept;
 };
+
+/// Compute the maximum magniture of the characteristic speeds.
```
magnitude

nilsdeppe committed:
Add doc-check target to check Doxygen for warnings

Closes #255

nilsdeppe commented:
Pushed a fixup commit, silenced a Doxygen warning in the ComputeDuDt commit, and added a commit that fixes #255 because I wanted to be able to easily find the warnings in the Doxygen spew.

kidder commented on src/Evolution/Systems/ScalarWave/Equations.hpp:
```diff
@@ -68,4 +70,111 @@ struct ComputeDuDt {
+ *                   - (n_i)_{\mathrm{ext}} F(\Pi)_{\mathrm{ext}}\right)
+ * \f}
+ * where \f$F^*\f$ is the normal dotted with the numerical flux and \f$F\f$ is
+ * the normal dotted with the flux, which computed in
```
which is computed

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,95 @@ void ComputeDuDt<Dim>::apply(
+  }
+}
+
+template <size_t Dim>
```
I got quite confused by what is stored in the packaged_data, please explicitly give the equation for what is stored by each tensor in the Vars in terms of the comment for the UpwindFlux struct.  And then you can remove the internal comment.

nilsdeppe commented:
I've squashed and rebased this on top of 0355860f (latest develop) while addressing @kidder's requests.

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+  // <NormalTimesFluxPi_i> = 0.5 * normal_dot_flux_pi * n_i
+  // <NormalDotFlux<Phi>_i> = 0.5 * normal_dot_flux_phi_i
+  //
+  // Note: There is a minus sign present now when sending to the neighbor
```
// Note: when Upwind::operator() is called, an Element passes in its own packaged data to fill the interior fields, and its neighbors packaged data to fill the exterior fields.  This introduces a sign flip for each normal used in computing the exterior fields.

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+}
+
+template <size_t Dim>
+void UpwindFlux<Dim>::operator()(
```
okay, I think to make this absolutely clear, the input arguments to this function should be named what they are:

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        phi_normal_dot_numerical_flux,
+    const gsl::not_null<Scalar<DataVector>*> psi_normal_dot_numerical_flux,
+    const Scalar<DataVector>& normal_dot_flux_pi_interior,
```
half_normal_dot_flux_pi_interior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+    const gsl::not_null<Scalar<DataVector>*> psi_normal_dot_numerical_flux,
+    const Scalar<DataVector>& normal_dot_flux_pi_interior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_phi_interior,
```
half_normal_dot_flux_phi_interior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+    const Scalar<DataVector>& normal_dot_flux_pi_interior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_phi_interior,
+    const Scalar<DataVector>& pi_interior,
```
half_pi_interior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+        normal_dot_flux_phi_interior,
+    const Scalar<DataVector>& pi_interior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_times_flux_pi_interior,
```
half_normal_times_flux_pi_interior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+    const Scalar<DataVector>& pi_interior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_times_flux_pi_interior,
+    const Scalar<DataVector>& normal_dot_flux_pi_exterior,
```
minus_half_normal_dot_flux_pi_exterior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+        normal_times_flux_pi_interior,
+    const Scalar<DataVector>& normal_dot_flux_pi_exterior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_phi_exterior,
```
minus_half_normal_dot_flux_phi_exterior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+    const Scalar<DataVector>& normal_dot_flux_pi_exterior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_dot_flux_phi_exterior,
+    const Scalar<DataVector>& pi_exterior,
```
half_pi_exterior

kidder commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,100 @@ void ComputeDuDt<Dim>::apply(
+        normal_dot_flux_phi_exterior,
+    const Scalar<DataVector>& pi_exterior,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        normal_times_flux_pi_exterior) const noexcept {
```
half_normal_times_flux_pi_exterior

nilsdeppe commented:
Okay, pushed fixes and squashed since they're pretty minor.

wthrowe commented on src/Evolution/Systems/ScalarWave/Equations.cpp:
```diff
@@ -27,6 +30,102 @@ void ComputeDuDt<Dim>::apply(
+  // <Pi> = 0.5 * pi
+  // <NormalDotFlux<Pi>> = 0.5 * normal_dot_flux_pi
+  // <NormalTimesFluxPi_i> = 0.5 * normal_dot_flux_pi * n_i
+  // <NormalDotFlux<Phi>_i> = 0.5 * normal_dot_flux_phi_i
```
Is doing the 0.5 factors this way actually gaining us anything?  It looks like we now do more additional multiplications in the package_data function than we save in operator() and the code is more confusing.  The only case where this could be more efficient is if the elements are stepping at different rates so we reuse the packaged data for multiple calls in the LTS algorithm, but I expect most boundaries will not be split like that.

I worry that this is a micro-optimization for the uncommon case that both slows down the common case and makes the code harder to understand.

nilsdeppe replied:
Alright, I did a benchmark(the 2, 4, etc. are the number of 1D grid points, the run is in 3D):

```
1/2 is in package_data
bench_upwind/2_mean           352 ns        352 ns    1994242
bench_upwind/2_median         350 ns        350 ns    1994242
bench_upwind/2_stddev           7 ns          7 ns    1994242
bench_upwind/4_mean           694 ns        694 ns    1005246
bench_upwind/4_median         690 ns        690 ns    1005246
bench_upwind/4_stddev          10 ns         10 ns    1005246
bench_upwind/6_mean          2708 ns       2708 ns     260929
bench_upwind/6_median        2699 ns       2700 ns     260929
bench_upwind/6_stddev          22 ns         22 ns     260929
bench_upwind/8_mean          6496 ns       6496 ns     107215
bench_upwind/8_median        6490 ns       6491 ns     107215
bench_upwind/8_stddev          49 ns         49 ns     107215
bench_upwind/10_mean        16767 ns      16768 ns      41589
bench_upwind/10_median      16753 ns      16753 ns      41589
bench_upwind/10_stddev        145 ns        145 ns      41589

1/2 is in call operator
bench_upwind/2_mean           282 ns        282 ns    2494796
bench_upwind/2_median         282 ns        282 ns    2494796
bench_upwind/2_stddev           4 ns          4 ns    2494796
bench_upwind/4_mean           583 ns        583 ns    1170752
bench_upwind/4_median         581 ns        581 ns    1170752
bench_upwind/4_stddev           6 ns          6 ns    1170752
bench_upwind/6_mean          2420 ns       2420 ns     291388
bench_upwind/6_median        2417 ns       2417 ns     291388
bench_upwind/6_stddev          20 ns         20 ns     291388
bench_upwind/8_mean          6089 ns       6089 ns     113148
bench_upwind/8_median        6081 ns       6082 ns     113148
bench_upwind/8_stddev          37 ns         37 ns     113148
bench_upwind/10_mean        15252 ns      15252 ns      45742
bench_upwind/10_median      15248 ns      15249 ns      45742
bench_upwind/10_stddev        134 ns        134 ns      45742
```

The code is below. I use `ClobberMemory()` to emulate that these things will be generally happening out of order and on different CPUs, so caches aren't helpful. Based on this, it's probably worth changing the 0.5 to be in the call operator. I'll do this and push a fixup.

```cpp
static void bench_upwind(benchmark::State& state) {
  const size_t pts_1d = state.range(0);
  constexpr const size_t Dim = 3;
  const Index<Dim> extents(pts_1d);
  CoordinateMaps::AffineMap map1d(-1.0, 1.0, -1.0, 1.0);
  using Map3d = CoordinateMaps::ProductOf3Maps<CoordinateMaps::AffineMap,
                                               CoordinateMaps::AffineMap,
                                               CoordinateMaps::AffineMap>;
  CoordinateMap<Frame::Logical, Frame::Grid, Map3d> map(
      Map3d{map1d, map1d, map1d});

  Variables<typename ScalarWave::UpwindFlux<Dim>::package_tags> packaged_data_0(
      pow<Dim>(pts_1d), 8.7),
      packaged_data_1(pow<Dim>(pts_1d), 8.7);

  Scalar<DataVector>
      normal_dot_flux_pi(pow<Dim>(pts_1d), 8.7),
      normal_dot_flux_psi(pow<Dim>(pts_1d), -8.7),
      pi(pow<Dim>(pts_1d), 2.0);
  tnsr::i<DataVector, Dim, Frame::Inertial> normal_dot_flux_phi(
      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 8.59)),
      unit_normal(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 1.0));

  Scalar<DataVector> half_normal_dot_flux_pi_interior(pow<Dim>(pts_1d), 3.0),
      half_normal_dot_flux_pi_exterior(pow<Dim>(pts_1d), -3.1),
      half_pi_interior(pow<Dim>(pts_1d), 4.1),
      half_pi_exterior(pow<Dim>(pts_1d), -4.2),
      pi_normal_dot_numerical_flux(pow<Dim>(pts_1d), 0.0),
      psi_normal_dot_numerical_flux(pow<Dim>(pts_1d), 0.0);
  tnsr::i<DataVector, Dim, Frame::Inertial> half_normal_dot_flux_phi_interior(
      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 7.89)),
      half_normal_times_flux_pi_interior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi,
                                                                     -9.27)),
      half_normal_dot_flux_phi_exterior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 3.99)),
      half_normal_times_flux_pi_exterior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi,
                                                                     -2.59)),
      phi_normal_dot_numerical_flux(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 0.0));

  ScalarWave::UpwindFlux<Dim> flux_computer{};

  while (state.KeepRunning()) {
    benchmark::ClobberMemory();
    flux_computer.package_data(make_not_null(&packaged_data_0),
                               normal_dot_flux_pi, normal_dot_flux_phi,
                               normal_dot_flux_psi, pi, unit_normal);
    benchmark::ClobberMemory();
    flux_computer.package_data(make_not_null(&packaged_data_1),
                               normal_dot_flux_pi, normal_dot_flux_phi,
                               normal_dot_flux_psi, pi, unit_normal);
    benchmark::ClobberMemory();
    flux_computer(make_not_null(&pi_normal_dot_numerical_flux),
                  make_not_null(&phi_normal_dot_numerical_flux),
                  make_not_null(&psi_normal_dot_numerical_flux),
                  half_normal_dot_flux_pi_interior,
                  half_normal_dot_flux_phi_interior, half_pi_interior,
                  half_normal_times_flux_pi_interior,
                  half_normal_dot_flux_pi_exterior,
                  half_normal_dot_flux_phi_exterior, half_pi_exterior,
                  half_normal_times_flux_pi_exterior);
    benchmark::ClobberMemory();
    flux_computer(make_not_null(&pi_normal_dot_numerical_flux),
                  make_not_null(&phi_normal_dot_numerical_flux),
                  make_not_null(&psi_normal_dot_numerical_flux),
                  half_normal_dot_flux_pi_interior,
                  half_normal_dot_flux_phi_interior, half_pi_interior,
                  half_normal_times_flux_pi_interior,
                  half_normal_dot_flux_pi_exterior,
                  half_normal_dot_flux_phi_exterior, half_pi_exterior,
                  half_normal_times_flux_pi_exterior);
  }
}
BENCHMARK(bench_upwind)->Arg(2)->Arg(4)->Arg(6)->Arg(8)->Arg(10);
```

kidder reviewed (commented):
I am happy for you to squash this
